
## Genetic analysis


source("R/packages.R")

## help here : https://www.molecularecologist.com/2016/02/26/quick-and-dirty-tree-building-in-r/
# and here https://fuzzyatelin.github.io/bioanth-stats/module-24/module-24.html

gen_data <- read.dna(here ("data","fas_gen_data.fas"), format = "fasta")
ind_sequenciados <- gsub(" ","", attr (gen_data,"dimnames")[[1]])

#------------------------------- #
#            MORPHOLOGY          
#------------------------------- #
## seleciona com base na presenca nos dados de traits
## atributos dos individuos
ind_trait <- read.csv (here ("data",'ind_trait.csv'),h=T,sep=";")

## selecionar as esp?cies de sigmodontine
sigmodontine <- unique (ind_trait$species)[-c(1,4,9,10,16)]

## pegar somente os atributos das sp de sigmodontine
ind_trait <- ind_trait[which (ind_trait$species %in% sigmodontine),]

# subset
gen_data  <- gen_data [which (ind_sequenciados %in% as.character(ind_trait$individuo) == T),]

# DNA INTO PHYDAT
gen_rodents <- phyDat(gen_data, type = "DNA", levels = NULL)
gen_rodents <- subset (gen_rodents, attr (gen_data,"dimnames")[[1]] %in% ind_trait$individuo)
# class(gen_rodents)

## NJ tree
gen_rodents_dnabin <- as.DNAbin (gen_rodents)
# genetic distance
dist_gen <- dist.dna(gen_rodents_dnabin, model="K80") # distance derived by Kimura (1980),
rodents_NJ  <- nj(dist_gen)

## bootstrapping the tree
myBoots <- boot.phylo(rodents_NJ, gen_rodents_dnabin, function(e) 
  root(nj(dist.dna(e, model = "K80")),1))

plot(rodents_NJ, show.tip=FALSE, edge.width=2)
title("NJ tree + bootstrap values")
tiplabels(frame="none", pch=20, col=rgb(0,0,0.1,alpha=0.1), cex=3, fg="transparent")
axisPhylo()
nodelabels(myBoots, cex=.6)

gen_rodents <- phyDat(gen_data, type = "DNA", levels = NULL)
gen_rodents <- subset (gen_rodents, attr (gen_data,"dimnames")[[1]] %in% ind_trait$individuo)
mt <- modelTest(gen_rodents)
#save(mt,file = here ("output", "test_model.RData"))

load(here ("output", "test_model.RData"))
env <- attr(mt, "env")
ls(envir=env)

round(mt$AICw,4)

## test diff models

(fitGTR <- eval(get("GTR+G+I", env), env))

## get the best and optimize tree
fitGTRopt <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                       rearrangement = "stochastic", control = pml.control(trace = 0))

# bootstrap to have branch support
nsamples<- 10 # 10000
bs.GTR <- bootstrap.pml(fitGTRopt, bs= nsamples, optNni=TRUE,
                        control = pml.control(trace = 0),
                        optGamma=TRUE, optInv=TRUE, model="GTR")


## cons tree
#cons <- consensus(bs.GTR, p = 0.7, check.labels = TRUE)

# plot opt tree and branch support
pdf(here("output","phyGTRFan.pdf"))
par(mfrow=c(1,1))
plotBS (midpoint(fitGTRopt$tree),
        bs.GTR,
        p = 5, 
        type="fan",
        cex=0.5)
axisPhylo(side=2)

dev.off()

pdf(here("output","phyGTR.pdf"))
plotBS (midpoint(fitGTRopt$tree),
        bs.GTR,
        p = 5, 
        type="p",
        cex=0.5,
        las=2)
axisPhylo(side=1)
dev.off()


#par(mfrow=c(2,1))
#par(mar=c(4,4,4,4))
#plotBS(midpoint(fitGTR$tree), bs.GTR, p = 90, type="p")
#title("a)")
#cnet <- consensusNet(bs.GTR, p=0.2)
#plot(cnet, "2D", show.edge.label=TRUE)
#title("b)")

# comparing a phylogeny without evol model,
# with one produced the diff transitions
comparePhylo(rodents_NJ, fitGTRopt$tree, 
             plot = TRUE, force.rooted = TRUE)

## depth of correction
dist_mat <- as.matrix(dist.p(gen_rodents, 
                             cost = "polymorphism", 
                             ignore.indels = TRUE),diag=T)

## mantel test testing p-distance and the distance generated by the GTR model 

mantel(dist_mat,
       as.dist(cophenetic(fitGTRopt$tree),diag=T))

## checking correspondence between distances
x <- as.vector(as.dist(cophenetic(rodents_NJ)))
y <- as.vector(as.dist(cophenetic(fitGTRopt$tree)))
plot(x, y, xlab="Pairwise distances on NJ tree", 
     ylab="Pairwise distances on the optimeed tree (GTR model)", 
     main="Is NJ appropriate?", 
     pch=20,  cex=1,col="gray")
abline(lm(y~x), col="red",lwd=2)
cor(y,x)^2


